### PR TITLE
Enum Class: Multiple Enums (itemdat.cpp)

### PR DESCRIPTION
**Pull request by kphoenix137**
Implements enum class to replace several item related enums:
`item_equip_type` -> `ItemEquipType`
`item_class` -> `ItemClass`
`item_cursor_graphic` -> `ItemCursorGraphic`
`item_misc_id` -> `ItemMiscID`
`_item_indexes` -> `ItemIndex`
`item_drop_rate` -> `ItemDropRate`
`unique_base_item` -> `UniqueBaseItem`
`item_effect_type` -> `ItemEffectType`
`goodorevil` -> `GoodOrEvil`
`ItemAnimationID`

Missing enumerators or enumerators lacking proper names have been added/updated.

Replaced `-1` with the respective proper enumerator in some spots in the code.

Replaced many numbers with enumerators.

Spells and item cursor graphics as item powers on Unique items have been replaced with enumerators.

### CODE REVIEW

**Comment by FitzRoyX** [comment]
Didn't you have another PR where you just fixed the names? I like the uppercase c style enums better.

**Comment by kphoenix137** [comment]
> Didn't you have another PR where you just fixed the names? I like the uppercase c style enums better.

These ones are safer to use since you can't accidently use the wrong type

**Comment by AJenbo** [comment]
I find them more readable and there is much better type checking then C style enums. About the only down side I see is that it's harder to use them as indexes as they have to be cast explicitly.

File: test/inv_test.cpp [hunk 1]
```diff
@@ -251,11 +251,11 @@ TEST_F(InvTest, RemoveCurrentSpellScroll_belt)
 		MyPlayer->SpdList[i].clear();
 	}
 	// Put a firebolt scroll into the belt
-	MyPlayer->executedSpell.spellId = static_cast<spell_id>(SPL_FIREBOLT);
+	MyPlayer->executedSpell.spellId = static_cast<SpellID>(SpellID::Firebolt);
```
Comment (AJenbo) [inline test/inv_test.cpp:254]: I don't think there is a need for this cast?

File: test/inv_test.cpp [hunk 2]
```diff
@@ -230,11 +230,11 @@ TEST_F(InvTest, RemoveCurrentSpellScrollFromInventoryFirstMatch)
 
 	// Put a firebolt scroll into the inventory
 	MyPlayer->_pNumInv = 1;
-	MyPlayer->executedSpell.spellId = static_cast<spell_id>(SPL_FIREBOLT);
+	MyPlayer->executedSpell.spellId = static_cast<SpellID>(SpellID::Firebolt);
```
Comment (AJenbo) [inline test/inv_test.cpp:233]:
```suggestion
	MyPlayer->executedSpell.spellId = SpellID::Firebolt;

```

File: Source/spelldat.h [hunk 3]
```diff
@@ -21,63 +21,63 @@ enum spell_type : uint8_t {
...
+	Warp,
+	Reflect,
+	Berserk,
+	RingFire,
+	Search,
+	RuneFire,
+	RuneLightning,
+	GreaterRuneLightning,
+	GreaterRuneFire,
+	RuneStone,
```
Comment (AJenbo) [inline Source/spelldat.h:77]: should be "RuneOf" instead of "Rune"

File: test/inv_test.cpp [hunk 4]
```diff
@@ -271,11 +271,11 @@ TEST_F(InvTest, RemoveCurrentSpellScrollFirstMatchFromBelt)
 		MyPlayer->SpdList[i].clear();
 	}
 	// Put a firebolt scroll into the belt
-	MyPlayer->executedSpell.spellId = static_cast<spell_id>(SPL_FIREBOLT);
+	MyPlayer->executedSpell.spellId = static_cast<SpellID>(SpellID::Firebolt);
```
Comment (AJenbo) [inline test/inv_test.cpp:274]:
```suggestion
	MyPlayer->executedSpell.spellId = SpellID::Firebolt;

```

File: test/inv_test.cpp [hunk 5]
```diff
@@ -212,11 +212,11 @@ TEST_F(InvTest, RemoveCurrentSpellScrollFromInventory)
 
 	// Put a firebolt scroll into the inventory
 	MyPlayer->_pNumInv = 1;
-	MyPlayer->executedSpell.spellId = static_cast<spell_id>(SPL_FIREBOLT);
+	MyPlayer->executedSpell.spellId = static_cast<SpellID>(SpellID::Firebolt);
```
Comment (AJenbo) [inline test/inv_test.cpp:215]:
```suggestion
	MyPlayer->executedSpell.spellId = SpellID::Firebolt;

```

File: test/inv_test.cpp [hunk 6]
```diff
@@ -87,21 +87,21 @@ TEST_F(InvTest, UseScroll_from_belt_invalid_conditions)
...
+	set_up_scroll(MyPlayer->SpdList[2], SpellID::Firebolt);
 	leveltype = DTYPE_TOWN;
-	EXPECT_FALSE(CanUseScroll(*MyPlayer, SPL_FIREBOLT));
+	EXPECT_FALSE(CanUseScroll(*MyPlayer, SpellID::Firebolt));
 
-	set_up_scroll(MyPlayer->SpdList[2], SPL_FIREBOLT);
-	MyPlayer->_pRSpell = static_cast<spell_id>(SPL_HEAL);
-	EXPECT_FALSE(CanUseScroll(*MyPlayer, SPL_HEAL));
+	set_up_scroll(MyPlayer->SpdList[2], SpellID::Firebolt);
+	MyPlayer->_pRSpell = static_cast<SpellID>(SpellID::Healing);
```
Comment (AJenbo) [inline test/inv_test.cpp:95]:
```suggestion
	MyPlayer->_pRSpell = SpellID::Healing;
```

File: test/inv_test.cpp [hunk 7]
```diff
@@ -64,21 +64,21 @@ TEST_F(InvTest, UseScroll_from_inventory_invalid_conditions)
...
+	set_up_scroll(MyPlayer->InvList[2], SpellID::Firebolt);
 	leveltype = DTYPE_TOWN;
-	EXPECT_FALSE(CanUseScroll(*MyPlayer, SPL_FIREBOLT));
+	EXPECT_FALSE(CanUseScroll(*MyPlayer, SpellID::Firebolt));
 
-	set_up_scroll(MyPlayer->InvList[2], SPL_FIREBOLT);
-	MyPlayer->_pRSpell = static_cast<spell_id>(SPL_HEAL);
-	EXPECT_FALSE(CanUseScroll(*MyPlayer, SPL_HEAL));
+	set_up_scroll(MyPlayer->InvList[2], SpellID::Firebolt);
+	MyPlayer->_pRSpell = static_cast<SpellID>(SpellID::Healing);
```
Comment (AJenbo) [inline test/inv_test.cpp:72]:
```suggestion
	MyPlayer->_pRSpell = SpellID::Healing;
```

File: Source/controls/touch/renderers.cpp [hunk 8]
```diff
@@ -107,17 +107,17 @@ void LoadButtonArt(ButtonTexture *buttonArt, SDL_Renderer *renderer)
...
+		ItemCursorGraphic::PotionMana,
+		ItemCursorGraphic::PotionRejuvenation,
+		ItemCursorGraphic::PotionFullHealing,
+		ItemCursorGraphic::PotionFullMana,
+		ItemCursorGraphic::PotionFullRejuvenation,
+		ItemCursorGraphic::Scroll
 	};
 
-	int potionFrame = static_cast<int>(CURSOR_FIRSTITEM) + static_cast<int>(ICURS_POTION_OF_HEALING);
+	int potionFrame = static_cast<int>(CURSOR_FIRSTITEM) + static_cast<int>(ItemCursorGraphic::PotionHealing);
```
Comment (AJenbo) [inline Source/controls/touch/renderers.cpp:120]:
```suggestion
	int potionFrame = static_cast<int>(CURSOR_FIRSTITEM) + static_cast<int>(ItemCursorGraphic::PotionOfHealing);
```

File: Source/debug.cpp [hunk 9]
```diff
@@ -556,9 +556,9 @@ std::string DebugCmdMinStats(const string_view parameter)
 std::string DebugCmdSetSpellsLevel(const string_view parameter)
 {
 	int level = std::max(0, atoi(parameter.data()));
-	for (int i = SPL_FIREBOLT; i < MAX_SPELLS; i++) {
-		if (GetSpellBookLevel((spell_id)i) != -1) {
-			SetSpellLevelCheat((spell_id)i, level);
+	for (int i = static_cast<int8_t>(SpellID::Firebolt); i < MAX_SPELLS; i++) {
+		if (GetSpellBookLevel((SpellID)i) != -1) {
+			SetSpellLevelCheat((SpellID)i, level);
```
Comment (AJenbo) [inline Source/debug.cpp:561]: Change to C++ `static_cast`

File: Source/items.h [hunk 10]
```diff
@@ -284,12 +284,12 @@ struct Item {
...
-		case ILOC_HELM:
-		case ILOC_ONEHAND:
-		case ILOC_RING:
-		case ILOC_TWOHAND:
+		case ItemEquipType::Amulet:
+		case ItemEquipType::Armor:
+		case ItemEquipType::Helm:
+		case ItemEquipType::OneHand:
+		case ItemEquipType::Ring:
+		case ItemEquipType::TwoHand:
```
Comment (AJenbo) [inline Source/items.h:292]:
```suggestion
		case ItemEquipType::TwoHands:
```

File: Source/towners.cpp [hunk 11]
```diff
@@ -401,7 +401,7 @@ void TalkToBlackSmith(Player &player, Towner &blackSmith)
 			return;
 		}
 
-		if (Quests[Q_ANVIL]._qvar2 == 1 && RemoveInventoryItemById(player, IDI_ANVIL)) {
+		if (Quests[Q_ANVIL]._qvar2 == 1 && RemoveInventoryItemById(player, ItemIndex::AnvilFury)) {
```
Comment (AJenbo) [inline Source/towners.cpp:404]:
```suggestion
		if (Quests[Q_ANVIL]._qvar2 == 1 && RemoveInventoryItemById(player, ItemIndex::AnvilOfFury)) {
```

File: Source/spelldat.h (hunk 3 above)
Comment (kphoenix137) [reply Source/spelldat.h:77]: Applied same principle to all things where "Of" was omitted

**Comment by kphoenix137** [comment]
Somehow spelldat.cpp got completely overwritten in the rebase, ill fix it tonight

**Comment by glebm** [comment]
Can you please separate out the `SpellID` rename?

**Comment by AJenbo** [comment]
image

**Comment by kphoenix137** [comment]
> Can you please separate out the `SpellID` one?

Created a new PR for that. If that new one gets merged, then I won't have to remove the changes from this one when I rebase.

**Comment by galaxyhaxz** [comment]
> I find them more readable and there is much better type checking then C style enums.

```
item_misc_id mid;

mid = MONSTER_FALLEN
...
Error: type is not valid
```

**Comment by kphoenix137** [comment]
> > I find them more readable and there is much better type checking then C style enums.
> 
> ```
> item_misc_id mid;
> 
> mid = MONSTER_FALLEN
> ...
> Error: type is not valid
> ```
> ```
> ItemMiscID MiscellaneousIdentity;
> MiscellaneousIdentity = ItemMiscID::pleaseReallyReallyLongSpelledOutNamesBecauseIDontUnderstandCommonAbbreviations;
> ```

I think if I wrote a piece of music and Beethoven insulted it, I wouldn't know whether to feel devastated or flattered. This is quite similar

**Comment by AJenbo** [comment]
> MiscellaneousIdentity = ItemMiscID::pleaseReallyReallyLongSpelledOutNamesBecauseIDontUnderstandCommonAbbreviations;

The enum value should start with upper case:
`pleaseReallyReallyLongSpelledOutNamesBecauseIDontUnderstandCommonAbbreviations`
->
`PleaseReallyReallyLongSpelledOutNamesBecauseIDontUnderstandCommonAbbreviations`

**Comment by AJenbo** [comment]
An example of what short of bugs will be caught by using `enum class` over `enum`:
```cpp
item_special_effect flag = ISPL_FIRE_ARROWS;
if (flag == MFLAG_NOHEAL) {
	monster.flags |= MFLAG_NOHEAL
}
```

The names can also sometimes be shorter then with a regular enum since they do not need to be globally unique and we can omit the prefix when using them.

```cpp
enum class DamageType {Fire}
enum class Immunity {Fire}

void ShowDamage(DamageType type)
  using enum DamageType;
  switch (tyoe) {
    case Fire:
...
  }
}
```

**Comment by galaxyhaxz** [comment]
Yeah... there were definitely bugs like the infamous Vileness affix not working due to them using wrong enum comparison.

However like most of these changes, you are trading one thing for another. You get safety, at the cost of more nuanced casting/more typing. That's what the joke getting at :P

**Comment by AJenbo** [comment]
Yeah it's a trade of, mostly just giving context to anyone who isn't aware what the benefits might be. For things that are only used for indexing `enum class` does not make sens as you would need to type case it on every usage.

**Comment by glebm** [comment]
> ```c++
> using enum DamageType;
> ```

This is C++20 unfortunately.

For small enums one can do this instead:

https://github.com/diasurgical/devilutionX/blob/c534e0e93c07d6d6d85151eeccac1d5699b988da/Source/objdat.cpp#L162-L167

**Comment by kphoenix137** [comment]
I would like to apologize for yet making this PR even larger. Fortunately this enum class has far less overall reach

**Comment by AJenbo** [comment]
If you can break this out in to smaller PRs it would be much faster to review and merge them

**Pull request closed without merging**